### ISSUE TITLE
Loss meaning

### ISSUE THREAD

**Issue by ammaccabanane95** [tpark94/speedplusbaseline#7]
Hi! Thank you very much for showing us the work! I have a question. I implemented your Keypoint regression network inside my code. I got good results but I don't understand very well the "loss_x" and y values. This value is 30000 at the beginning of the train and reach after some epochs the value of 2000 more or less (the value inside the brackets, I think it is referred to "loss_x_meter.avg"). I don't understand if it is ok, because I expect lower values (of the order of 10 or less) since the keypoint calculation is successful after the train. Thank you for your attention.
Best regards,

Antonio

**Comment by tpark94** [comment]
That does sound weird because the keypoints that are used in the loss function should be normalized to [0, 1] within the RoI. 

I've also checked my training log and `loss_x` quickly decreases to below 1.

The only way that value (~2000) would make sense is if you are computing the loss function using pixel values instead of normalized ones.

Could you try print and see what the `xc`, `txc` values are before you compute the loss below?

```
  # Loss
  loss_x, loss_y = 0.0, 0.0
  for i in range(self.nK):
      loss_x = loss_x + self.loss(xc[:,i], txc[:,i])
      loss_y = loss_y + self.loss(yc[:,i], tyc[:,i])
  loss = loss_x + loss_y
```

Best,
Jeff

**Comment by ammaccabanane95** [comment]
Hi Jeff, thank you for the quick answer!

Here is the print of xc: 
tensor([[386.9261, 198.6891, 456.5170, 425.0045,  81.9933,  47.6443],
        [437.1342,  81.3943, 285.3909, 258.6887, 119.4660,  93.4669],
        [259.6613, 265.8788,  43.4623,  47.1550, 475.1259, 477.8344],
        [236.1028, 273.8059,  31.9900,  43.8793, 472.0091, 485.3218],
        [ 53.3810, 455.4149, 354.2054, 403.8903, 260.7128, 309.7231],
        [134.0004, 308.9504, 416.9582, 444.3201,  72.8022,  99.6672],
        [429.2105,  62.0907, 156.7779, 105.7900, 248.1820, 200.4458],
        [399.4531, 200.9363, 488.3024, 460.1680,  59.4438,  30.5994],
        [444.4197,  74.4478, 293.8248, 241.9660, 146.5123,  90.9556],
        [153.3568, 292.4175, 463.4783, 482.7593,  29.2396,  47.7609],
        [298.8038, 255.5355,  95.1396,  90.7854, 445.2635, 437.7912],
        [ 55.2616, 385.5528,  91.1054, 145.7960, 415.9376, 470.8224],
        [479.1827,  51.2884, 147.7961,  94.0183, 270.2878, 217.5712],
        [442.2416, 102.4922, 144.3892,  94.3890, 317.2864, 266.5401],
        [267.4326, 252.4600, 442.8120, 438.5170,  76.8568,  70.3889],
        [ 81.2696, 375.3776, 145.4361, 194.6845, 369.6159, 419.1401],
        [335.3913, 224.5027, 470.2369, 455.1230,  58.1409,  42.2348],
        [316.5754, 228.5594, 445.3228, 437.5687,  72.7732,  65.9797],
        [455.5263,  67.1189, 142.5921,  95.7673, 270.3112, 224.0726],
        [254.0782, 255.7898,  47.3421,  45.1415, 468.9619, 464.1078],
        [272.3385, 230.4912, 247.1927, 262.5351, 208.8540, 228.9897],
        [178.9609, 300.1054,  54.4484,  71.3583, 444.6592, 461.9102],
        [101.8770, 420.6581, 291.7318, 326.7394, 313.1667, 348.7959],
        [130.4283, 324.3138, 437.5203, 467.0345,  54.1228,  80.4513],
        [466.1605,  53.8383, 114.6009,  60.5832, 296.4403, 240.4300],
        [196.5933, 288.0373, 120.2460, 134.1543, 387.3921, 401.3824],
        [ 33.1185, 454.2923, 352.6199, 396.9758, 263.4353, 308.8909],
        [455.4199,  64.5204, 225.0557, 170.3319, 201.2866, 146.9484],
        [ 42.4196, 425.6219, 121.7757, 171.4390, 432.9379, 483.9283],
        [ 93.6651, 418.3922, 227.9498, 270.3625, 368.0189, 409.7760],
        [346.9145, 240.6514, 130.1312, 102.7438, 446.2060, 416.0851],
        [167.6477, 299.6910, 125.6396, 123.6401, 405.6583, 407.9014]],
       device='cuda:0', grad_fn=<IndexBackward0>)



and txc:
tensor([[382.5785, 213.4420, 464.3655, 441.2166,  69.2960,  48.8593],
        [438.2977,  75.2881, 281.2381, 232.8792, 146.5588, 100.0599],
        [265.6921, 265.7331,  35.3665,  41.8802, 471.3809, 478.0266],
        [247.5949, 272.4618,  34.7628,  42.7474, 466.9692, 478.0565],
        [ 58.3389, 455.1329, 368.1144, 412.0475, 253.3237, 300.9627],
        [133.8213, 301.3404, 413.4379, 453.0655,  61.0614, 102.3202],
        [447.1547,  65.3287, 168.4901,  90.7082, 257.9908, 192.6248],
        [406.7354, 201.8877, 475.7626, 452.7681,  63.6899,  37.1801],
        [447.1100,  69.8486, 303.4727, 237.7715, 140.3278,  67.7799],
        [144.1758, 279.8016, 466.4850, 486.4854,  26.1794,  51.9197],
        [291.2801, 266.3125,  67.5894,  92.0009, 426.8481, 446.8104],
        [ 59.3044, 372.7981, 121.7477, 138.6797, 430.8233, 454.8687],
        [454.4597,  58.3581, 163.4323,  95.2879, 263.9436, 202.3909],
        [436.6469,  94.1563, 114.7552,  77.8903, 330.3928, 294.3068],
        [263.8719, 261.1019, 448.7739, 423.8632,  86.6455,  64.9479],
        [ 81.0597, 383.6709, 139.5863, 204.0311, 364.7828, 432.4753],
        [343.0688, 223.6408, 473.9872, 456.0316,  55.2934,  38.3858],
        [298.6492, 218.6587, 456.6335, 428.5988,  81.3041,  55.7962],
        [446.8470,  67.6196, 146.1971,  83.8604, 274.3657, 219.0630],
        [264.2264, 254.9390,  41.5630,  40.2256, 473.0394, 468.6096],
        [298.8553, 233.2057, 472.8840, 466.6840,  52.8058,  41.0590],
        [183.1069, 309.7785,  57.4601,  58.1887, 455.9310, 457.4586],
        [ 95.8080, 419.2810, 326.1720, 337.0099, 317.9588, 328.7935],
        [120.1582, 312.1152, 448.3440, 472.2383,  39.8433,  66.8323],
        [470.6656,  46.1450,  93.0105,  43.4462, 278.5117, 234.5995],
        [188.6626, 296.7961, 103.5518, 152.9373, 361.3849, 412.6632],
        [ 52.4156, 461.6995, 360.1570, 401.8937, 262.8413, 312.1027],
        [456.9454,  57.0359, 210.9637, 167.2959, 200.8512, 152.9010],
        [ 40.8021, 392.6971, 131.5611, 174.0883, 425.8834, 472.1918],
        [104.4825, 410.1374, 241.1275, 262.3886, 380.5612, 401.0759],
        [325.2424, 208.6261,  94.8790,  60.6181, 453.0853, 412.3421],
        [164.3131, 299.7179, 121.6323, 104.7432, 411.8057, 392.3519]],
       device='cuda:0')

As you can see my batch size is 32 and the dimension of my images is 512x512.
Thank you again,
Have a nice day

Antonio

**Comment by tpark94** [comment]
Thanks, as I suspected, you are using pixel values of keypoints to compute the loss, instead of normalized values.

For example, this normalization of ground-truth keypoint locations is done in `src/datasets/transforms.py`, line 156:
```
# Adjust keypoints (0 ~ 1)
keypts = torch.tensor(keypts, dtype=torch.float32)
keypts[0] = (keypts[0] - xmin) / (xmax - xmin)
keypts[1] = (keypts[1] - ymin) / (ymax - ymin)
```
but I assume that you are not doing this. Either way, the loss value (~2000) makes sense for pixels.

1. The loss is `MSELoss`, which computes mean squared error, so if your prediction of one keypoint is off by 10 pixels, the loss would be ~100.
2. As you see in my previous post, the loss is being summed for 11 keypoints, which would make your total loss ~1100. 

From the values of `xc` and `txc` you've provided, I see differences well over 10 pixels in many cases.

Jeff

**Comment by ammaccabanane95** [comment]
Thank you very much again!
Best,

Antonio
